lib/types.nix

```
@@ -710,6 +710,23 @@ rec {
         in mergedOption.type;
     };
 
+    # A type that is one of several submodules, similiar to types.oneOf but is usable inside attrsOf or listOf
+    # submodules need an option with a type str which is used to find the corresponding type
+    taggedSubmodules =
+      { types
+      , specialArgs ? {}
+      }: mkOptionType rec {
+      name = "taggedSubmodules";
+      description = "one of ${concatStringsSep "," (attrNames types)}";
```

```suggestion
      description = "submodule with type tag";
      descriptionClass = "composite";
```

The valid type values can be documented in a generated `type` option.